syntax = "proto3";

package network;

message init_request {
    uint32 key = 1;
}

message init_response {
    uint64 beaver_length = 1;

    message node {
        string ip = 1;
        uint32 port = 2;
    }

    repeated node nodes = 2;
}

message node_request {
    uint32 key = 1;
    bytes input = 2;

    enum Mode {
        PLAIN = 0;
        AES = 1;
    }
    Mode mode = 3;

    bytes ai = 4;
    bytes bi = 5;
    bytes ci = 6;

    bytes aes_key = 7;
    bytes aes_iv = 8;
    int32 length = 9;
}

message node_response {
    uint32 result = 1;
}

message node_register_response {
    uint32 index = 1;
    bytes dfa = 2;
}

message node_notify {
    uint32 key = 1;
    uint32 turn = 2;
    bytes xa = 3;
    bytes yb = 4;
}

message server_broadcast {
    uint32 key = 1;
    uint32 turn = 2;
    bytes xa = 3;
    bytes yb = 4;
}

message general_response {
    enum Status {
        SUCCESS = 0;
        FAIL = 1;
    }
    Status status = 1;
}